### PR TITLE
Add fixed-width variant of ECSV

### LINKED ISSUE #386: Add option to input/output column units for fixed width tables
Extend the `io.ascii.FixedWidth` reader to include a keyword arg that will specify that there is a row of unit specifiers after the column name specifiers (or at the top of the header if there are no column names).  This will apply for both reading and writing fixed width tables.

This allows for outputting a table to a file in a format like `Table.pprint` with `show_units=True`, and then reading back that table with no information loss.

### LINKED ISSUE #12803: Human readable ecsv table
I find the ecsv extremely useful, I'm a great fan of including units, comments and descriptions in the tables. The only downside of this format is that I can't obtain a ecsv (with a comma as a delimiter, since it guarantees max compatibility) that has all the commas aligned. I realize that it is not in the specs of the format to have all the commas aligned, but researchers that I know,  that are interested in using tables in astropy,  are not adopting it because they can't read the table with their eyes from the ASCII file. So they try to use fixed-size tables, which produce issue with descriptions and units, leading them to abandon any effort to switch to astropy tables. 

I tried to experiment, and a MWE is: 

```import numpy as np
from astropy.table import Table

a = np.array([1.20, 1.21, 2.22])
b = np.array(['u', 'g', 'egr'])
c = np.array([4.20, 1.00, 2.20])

t = Table()
t['a'] = a
t['b'] = b
t['c'] = c

t.write('tmp.txt', format='ascii.ecsv', delimiter=',', strip_whitespace=False,  overwrite=True)
```

resulting in
```# %ECSV 1.0
# ---
# datatype:
# - {name: a, datatype: float64}
# - {name: b, datatype: string}
# - {name: c, datatype: float64}
# delimiter: ','
# schema: astropy-2.0
a,b,c
1.2,u,4.2
1.21,g,1.0
2.22,egr,2.2
```

I modified the file `astropy/astropy/io/ascii/ecsv.py` of the tag v5.0.1, line 383, from `return str(col[idx])` to `return str(col[idx]).rjust(4)` and as an output I get

```
# %ECSV 1.0
# ---
# datatype:
# - {name: a, datatype: float64}
# - {name: b, datatype: string}
# - {name: c, datatype: float64}
# delimiter: ','
# schema: astropy-2.0
a,b,c
 1.2,   u, 4.2
1.21,   g, 1.0
2.22, egr, 2.2
```
which is what I wanted. Of course this is not the way to go, and probably I'm breaking also something somewhere else, but would it be possible to obtain such a feature? My operating system is High Sierra, python 3.9.7 intel, numpy 1.21.2, astropy 5.0.1.

### PR DESCRIPTION
**Pull request by taldcroft**
### Description

This pull request is to make a "human readable" variant of ECSV that adds spaces to the data values so that they are aligned in columns (i.e. fixed width). This is done as a new format to make it clear that this is not strictly ECSV although the output is readable by the Python ECSV reader for both space and comma-delimited outputs. 

The behavior of non-astropy CSV readers with the extra spaces is not guaranteed.

I'd like to say this can close the issue of including table meta data (units in particular) in fixed width tables, starting from #386 almost 10 years ago.

To do:
- [x] Tests
- [x] Docs
- [x] Change log
- [x] What's new

```
>>> t = Table()
>>> t['col0'] = [1, 2, 30000, -5]
>>> t['col1'] = ['x z', 'y,z', 'z, x', 'hello!']
>>> t['col2'] = [1.0, -20000.0, 3.0, 4.0]

>>> ascii.write(t, format='ecsv_fixed_width')
# %ECSV 1.0
# ---
# datatype:
# - {name: col0, datatype: int64}
# - {name: col1, datatype: string}
# - {name: col2, datatype: float64}
# schema: astropy-2.0
 col0    col1      col2
    1   "x z"       1.0
    2     y,z  -20000.0
30000  "z, x"       3.0
   -5  hello!       4.0

>>> ascii.write(t, format='ecsv_fixed_width', delimiter=',')
# %ECSV 1.0
# ---
# datatype:
# - {name: col0, datatype: int64}
# - {name: col1, datatype: string}
# - {name: col2, datatype: float64}
# delimiter: ','
# schema: astropy-2.0
 col0,   col1,     col2
    1,    x z,      1.0
    2,  "y,z", -20000.0
30000, "z, x",      3.0
   -5, hello!,      4.0
```

Fixes #12803
Fixes #386

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by taldcroft** [comment]
@NegriAndrea - does this solve your problem of human-readable ECSV?

@mhvk @hamoguo @astrofrog @eteq - blast from the past. How about this as a compromise for capturing units in a fixed-width table? Probably overtaken by time/events, but there was #2869  which generated a lot of discussion without any code getting merged.

**Comment by NegriAndrea** [comment]
@taldcroft - this solves exactly my problem, thanks! It's fantastic that you guys did it in such a short timescale, and I'm happy if my request fixes also the issue of having units on a fixed-format table.

**Comment by pllim** [comment]
I remilestoned this because it is still in draft form. FYI.

**Comment by taldcroft** [comment]
@mbtaylor - do you have any concerns about adding a fixed-width variant of ECSV (philosophically and practically)? Will such files read into TOPCAT?

After many discussions through the years of people wanting some sort of human-readable (fixed-width) tables that include units and possibly other meta-data, I thought a fixed-width ECSV might satisfy that need without inventing a whole new ASCII table format specification

**Comment by taldcroft** [comment]
@hamogu - well, this is pretty late but I think it's completely ready for review. :smile:

File: astropy/table/tests/test_mixin.py [hunk 1]
```diff
@@ -874,7 +874,6 @@ class MyArrayWrapper(ArrayWrapper):
 
     t = Table()
     t['tm'] = MyArrayWrapper([0, 1, 2])
-    out = StringIO()
```
Comment (taldcroft) [inline astropy/table/tests/test_mixin.py:877]: Unrelated clean-up.

File: astropy/time/core.py [hunk 2]
```diff
@@ -1106,7 +1106,7 @@ def _apply(self, method, *args, format=None, cls=None, **kwargs):
         new_format = self.format if format is None else format
 
         if callable(method):
-            apply_method = lambda array: method(array, *args, **kwargs)
+            apply_method = lambda array: method(array, *args, **kwargs)  # noqa: E731
```
Comment (taldcroft) [inline astropy/time/core.py:1109]: Unrelated clean-up.

**Comment by taldcroft** [comment]
Indeed things get slightly fuzzy with respect to the treatment of extra whitespace, but the convention of stripping leading whitespace after stripping by the delimiter is not uncommon even if it is not rigorous to the CSV spec RFC-4180 (https://datatracker.ietf.org/doc/html/rfc4180).

There is the Python CSV reader with https://docs.python.org/3/library/csv.html#csv.Dialect.skipinitialspace. Astropy io.ascii generally ignores both initial and trailing whitespace (for better or worse). Pandas is a bit inconsistent: if the leading/trailing whitespace is part of a field that would otherwise be numeric, then it is ignored, but if it is part of string field then it is kept:
```
In [16]: cat junk.dat
a  ,  b, c
 1 , 2.0  ,          hello world
2,3,jim

In [17]: df = pandas.read_csv('junk.dat')

In [18]: df
Out[18]: 
   a      b                      c
0    1  2.0            hello world
1    2  3.0                    jim

In [19]: df.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   a       2 non-null      int64  
 1     b     2 non-null      float64
 2    c      2 non-null      object 
dtypes: float64(1), int64(1), object(1)
memory usage: 176.0+ bytes

In [23]: repr(df[' c'][0])
Out[23]: "'          hello world'"

```

Anyway, this is probably going to be controversial so we can skip it for this release. Astropy users have been waiting about 10 years so maybe another 6 months is OK.

**Comment by mbtaylor** [comment]
As @hamogu says, and as noted in @taldcroft's original comment, the example files above are not conformant to the ECSV 1.0 standard as written, and for that reason the STIL/TOPCAT reader doesn't read them as presumably intended.  For the space-separated variant each padding space is taken as a delimiter, so there are far too many fields per row and the reader rejects the file.  For the comma-separated variant, the leading spaces are taken to be part of the value tokens, which leads to string-valued field values having leading spaces and in some cases numeric values failing to parse.

I'm quite happy to adjust the STIL/TOPCAT reader, and maybe writer, to work with an ECSV-like format that looks like these examples, but I would like to have the definition of the format, or at least the behaviour expected from a reader, written down explicitly somewhere rather than just showing a couple of examples and assuming that the intended interpretation is obvious.  Such definition could either be an update to the ECSV standard (my preferred option) or text in some external location.  In particular I'd like to be convinced that input files are not ambiguous, i.e. it's not the case that an 'enhanced' ECSV reader and a strict ECSV 1.0 reader could both read the same file and come up with different meanings.

**Comment by taldcroft** [comment]
Thanks @mbtaylor and @hamogu. All good points, so this will take either more work to make it solid or else go in the bin of "not-quite" ideas.

**Comment by saimn** [comment]
@taldcroft - So I guess it can be moved to 5.2 ?

**Comment by taldcroft** [comment]
@saimn - I just cleared the milestone, not sure now if this is even a good idea...

**Comment by NegriAndrea** [comment]
I am in no way expert on this, but I would like to give my contribution as well. As far as I understand the ECSV mandates that every column has a type, which can be bool, different flavors of int and float, and string. With a padding character a blank space, and a separator a comma, if a column is a bool or numerical, one could strip the leading padding characters, since we know it's a bool or a number (correct me if I'm wrong). The issue comes with the string type. What if we do not allow the string to contain the padding character? Would that work?

**Comment by taldcroft** [comment]
As mentioned earlier, maybe this was a half-baked idea. Closing.
